packages/engine/Source/Shaders/Model/AtmosphereStageFS.glsl

```diff
@@ -63,49 +64,139 @@ void applyFog(inout vec4 color, vec4 groundAtmosphereColor, vec3 lightDirection,
+        vec3 radii = czm_ellipsoidRadii;
+        float maxRadius = max(radii.x, max(radii.y, radii.z));
+        fadeRange -= maxRadius;
+    }
+    return fadeRange;
+}
+
+void applyGroundAtmosphere(inout vec4 color, vec4 groundAtmosphereColor, vec3 positionWC, vec3 atmosphereLightDirectionWC) {
+    float cameraHeight = getCameraHeight();
+
+    // If the camera is inside the earth, skip rendering the ground atmosphere
+    vec3 radii = czm_ellipsoidRadii;
+    float minRadius = min(radii.x, min(radii.y, radii.z));
+    if (cameraHeight < minRadius) {
+        return;
+    }
+
+    // Approximate normal at this point on the globe.
+    vec3 normalWC = normalize(positionWC);
+
+    // Blend between constant lighting when the camera is near the earth
+    // and Lambert diffuse shading + ground atmosphere when the camera is farther away
+    vec2 fadeRange = getLightingFadeRange();
+    float lightingFade = fade(cameraHeight, fadeRange, vec2(0.0, 1.0));
+
+    // Apply lambert diffuse shading based on the light
+    float lambert = czm_getLambertDiffuse(normalWC, atmosphereLightDirectionWC);
+    const float diffuseMultiplier = 5.0;
+    const float vertexShadowDarkness = 0.3;
+    float diffuseIntensity = clamp(lambert * diffuseMultiplier + vertexShadowDarkness, 0.0, 1.0);
+    diffuseIntensity = mix(1.0, diffuseIntensity, lightingFade);
+    vec4 diffuseColor = vec4(color.rgb * czm_lightColor * diffuseIntensity, color.a);
+
+    // The transmittance is based on optical depth i.e. the length of segment of the ray inside the atmosphere.
+    // This value is larger near the "circumference", as it is further away from the camera. We use it to
+    // brighten up that area of the ground atmosphere.
+    const float transmittanceModifier = 0.5;
+    float transmittance = transmittanceModifier + clamp(1.0 - groundAtmosphereColor.a, 0.0, 1.0);
+    vec3 diffuseAndGroundAtmosphere = diffuseColor.rgb + groundAtmosphereColor.rgb * transmittance;
+
+    vec3 finalAtmosphereColor = diffuseAndGroundAtmosphere;
+
+    if (czm_atmosphereDynamicLighting != ATMOSPHERE_DYNAMIC_LIGHTING_NONE) {
+        // Use a dot product (similar to Lambert shading) to create a mask that's positive on the
+        // daytime side of the globe and 0 on the nighttime side of the globe.
+        //
+        // Use this to select the diffuse + ground atmosphere on the daytime side of the globe,
+        // and just the ground atmosphere (which is much darker) on the nighttime side.
+        float dayNightMask = clamp(dot(normalWC, atmosphereLightDirectionWC), 0.0, 1.0);
+        vec3 dayNightColor = mix(groundAtmosphereColor.rgb, diffuseAndGroundAtmosphere, dayNightMask);
+
+        // Fade in the day/night color in as the camera height increases towards space.
+        float nightFade = fade(cameraHeight, czm_atmosphereNightFadeRange, vec2(0.05, 1.0));
+        finalAtmosphereColor = mix(diffuseAndGroundAtmosphere, dayNightColor, nightFade);
```

Comment:
in `GlobeFS.glsl`, `u_nightFadeDistance`, I found some of the variable names and usage a bit backwards to my understanding, so I rephrased it and documented the conventions I'm using. 

It's mathematically equivalent, I just thought this way would be a little easier to understand.